bionic/code_analyzer.py

```diff
@@ -0,0 +1,134 @@
+"""
+This module implements the logic to analyze your code objects and
+returns the result in the form of a byte array.
+"""
+
+import inspect
+import json
+import warnings
+
+from .utils.misc import oneline
+
+
+DEFAULT_NAME = b"zA)T+ndu/Ygd%JH=eux)[F9:<>)zND]}u4x,c49tWH<yd5+#+4Xg{=J<[S5/'XG+"
+
+
+def analyze_func(func):
+    assert inspect.isroutine(func)
+
+    analyzer = CodeAnalyzer()
+    analyzer.analyze(func)
+    return analyzer.get_analyzed_bytes()
+
+
+class CodeAnalyzer:
+    """
+    This class contains the logic to analyze a code object. For now, it
+    only analyzes the function's bytecode and constants. But in the
+    future, it will be able to analyze its references, including class
+    types.
+    """
+
+    def __init__(self):
+        self._id_to_hash = {}
+        self._ids_being_hashed = set()
+        self._analyzed_bytearray = bytearray()
+
+    def get_analyzed_bytes(self):
+        return self._analyzed_bytearray
+
+    def analyze(self, obj):
+        obj_id = get_obj_id(obj)
+
+        # If already hashed, return the hash.
+        if obj_id in self._id_to_hash:
+            return self._id_to_hash[obj_id]
+
+        # If the obj is already hashing, break the circular ref and return early.
+        if obj_id in self._ids_being_hashed:
+            return get_obj_name(obj, DEFAULT_NAME)
+        self._ids_being_hashed.add(obj_id)
+
+        # Otherwise, hash the object.
+        obj_hash = self._analyze(obj)
```

Comment:
What is happening here? Does `_analyze` return anything?

For that matter, should this function return something?

Reply:
This method converts the objects into bytes and adds them to the result. I added some documentation to this method as well.